### DISCUSSION TITLE
The question about the madnet.py

### DISCUSSION THREAD

**Discussion by Yu-Cheung** [ChristianOrr/Real-time-self-adaptive-deep-stereo#1] [Q&A] [answered]
Hello, thank for your code, it's very useful! When read the code in madnet.py, I find a question that I don't understand.

From the paper, I found the image has been scaled, but it doesn't seem to be scaled in the code, is that right?

```# Initializing the layers
    layer_kwargs = {
        "kernel_size": (3, 3),
        "padding": "same",
        "activation": tf.keras.layers.Activation(tf.nn.leaky_relu, dtype=tf.float32, name="leaky_relu"),
        "use_bias": True
    }
    # Image feature pyramid (feature extractor)
    # F1
    conv1 = tf.keras.layers.Conv2D(
        filters=16,
        strides=2,
        name="conv1",
        input_shape=(input_shape[0], input_shape[1], input_shape[2], ),
        **layer_kwargs)
    conv2 = tf.keras.layers.Conv2D(filters=16, strides=1, name="conv2", **layer_kwargs)
    # F2
    conv3 = tf.keras.layers.Conv2D(filters=32, strides=2, name="conv3", **layer_kwargs)
    conv4 = tf.keras.layers.Conv2D(filters=32, strides=1, name="conv4", **layer_kwargs)
    # F3
    conv5 = tf.keras.layers.Conv2D(filters=64, strides=2, name="conv5", **layer_kwargs)
    conv6 = tf.keras.layers.Conv2D(filters=64, strides=1, name="conv6", **layer_kwargs)
    # F4
    conv7 = tf.keras.layers.Conv2D(filters=96, strides=2, name="conv7", **layer_kwargs)
    conv8 = tf.keras.layers.Conv2D(filters=96, strides=1, name="conv8", **layer_kwargs)
    # F5
    conv9 = tf.keras.layers.Conv2D(filters=128, strides=2, name="conv9", **layer_kwargs)
    conv10 = tf.keras.layers.Conv2D(filters=128, strides=1, name="conv10", **layer_kwargs)
    # F6
    conv11 = tf.keras.layers.Conv2D(filters=192, strides=2, name="conv11", **layer_kwargs)
    conv12 = tf.keras.layers.Conv2D(filters=192, strides=1, name="conv12", **layer_kwargs)

    #############################SCALE 6#################################
    M6 = ModuleM(layer="6", search_range=search_range)
    ############################SCALE 5###################################
    M5 = ModuleM(layer="5", search_range=search_range)
    ############################SCALE 4###################################
    M4 = ModuleM(layer="4", search_range=search_range)
    ############################SCALE 3###################################
    M3 = ModuleM(layer="3", search_range=search_range)
    ############################SCALE 2###################################
    M2 = ModuleM(layer="2", search_range=search_range)```

**Comment by ChristianOrr** [comment]
Hi Asurasin,

I'm glad you found value in it! 

I did scale the images, but not in the model. The stereo images are scaled using the[ _get_image](https://github.com/ChristianOrr/madnet-deep-stereo-with-keras/blob/36cd6db182b7edafa0680b61e0836a8007a88622/preprocessing.py#LL64C18-L64C18) method in the [StereoDatasetCreator](https://github.com/ChristianOrr/madnet-deep-stereo-with-keras/blob/36cd6db182b7edafa0680b61e0836a8007a88622/preprocessing.py#L6) class.

**Reply by Yu-Cheung** [reply]
Hi ChristianOrr,
Thank you for your reply! 
I have seen, and I'm using Tensorflow for the first time, so I probably didn't understand the parameters of the conv layers.

**Reply by ChristianOrr** [reply, marked as answer]
No problem. I'm glad I could help. 

Scaling is also performed on the feature maps inside the model to achieve the pyramid structure. The resolution of the feature maps is decreased by setting the stride=2 in the convolutional layers. The resolution is then increased in the second half of the network by using bilinear interpolation in the [ModuleM](https://github.com/ChristianOrr/madnet-deep-stereo-with-keras/blob/36cd6db182b7edafa0680b61e0836a8007a88622/madnet.py#L284) class. It's also possible to increase the resolution of the feature maps by using transposed convolutions with stride=2, but those weren't used in this model.

**Reply by Yu-Cheung** [reply]
Thank you for your reply.
I understand. And I have another problem. In the **train_quant.py**, I found that you import a module named `tensorflow_model_opitimization`, and I can't find this module. What's the differences between the **train.py**, **train_quant.py** and the **train_wandb.py**? 
I found that you don't use the adaptive methods in the paper, and how add it into the code? Just like the code in the tf1.x method?
Thanks!

**Reply by ChristianOrr** [reply]
1. The train_quant.py is meant for quantization-aware training. I never got it working, but I didn't want to delete the code in case I decided to take another look. 
2. train.py is the default training script, and train_wandb.py is for training using Weights and Biases. This is useful if you want to perform hyperparameter sweeps.
3. I have already implemented the MAD adaptation method. MAD++ is not implemented. If you want to add MAD++ you need to create proxy ground truths. You can do this by extracting disparity using Semi-Global Matching, and then filtering out all the low-confidence disparities using a confidence measure. I have done an implementation of SGM in Numpy from scratch in this repo: [Semi-Global Matching Numpy](https://github.com/ChristianOrr/semi-global-matching-numpy), OpenCV can also do SGM. I've also implemented a few confidence measures in Numpy here: [Confidence Measures](https://github.com/ChristianOrr/confidence-measures). If you create these proxy groundtruth's and save them as pfm disparity files, then you can just point to the directory they're stored in when training using the [train_disp_dir](https://github.com/ChristianOrr/madnet-deep-stereo-with-keras/blob/36cd6db182b7edafa0680b61e0836a8007a88622/train.py#LL13C31-L13C31) argument. If you want to do online adaptation using MAD++, then the monkey patched predict step will need to be updated, [_custom_predict_step](https://github.com/ChristianOrr/madnet-deep-stereo-with-keras/blob/36cd6db182b7edafa0680b61e0836a8007a88622/madnet.py#L390). 

I'm going to close this discussion as we're going pretty far off the original question. You're welcome to open a new discussion to discuss any of these further.
